fetch/mesh_data.go

```diff
@@ -23,43 +23,73 @@ func (f *Fetch) GetAtxs(ctx context.Context, ids []types.ATXID) error {
 	if len(ids) == 0 {
 		return nil
 	}
-	f.logger.WithContext(ctx).With().Debug("requesting atxs from peer", log.Int("num_atxs", len(ids)))
+	f.logger.WithContext(ctx).With().Info("requesting atxs from peer", log.Int("num_atxs", len(ids)))
 	hashes := types.ATXIDsToHashes(ids)
-	return f.getHashes(ctx, hashes, datastore.ATXDB, f.validators.atx.HandleMessage)
+	return f.getHashes(ctx, hashes, datastore.ATXDB, f.validators.atx.HandleMessage, withLimiter(f.getAtxsLimiter))
 }
 
 type dataReceiver func(context.Context, types.Hash32, p2p.Peer, []byte) error
 
+type getHashesOpt func(*getHashesOpts)
+
+func withLimiter(l limiter) getHashesOpt {
+	return func(o *getHashesOpts) {
+		o.limiter = l
+	}
+}
+
 func (f *Fetch) getHashes(
 	ctx context.Context,
 	hashes []types.Hash32,
 	hint datastore.Hint,
 	receiver dataReceiver,
+	opts ...getHashesOpt,
 ) error {
+	options := getHashesOpts{
+		limiter: noLimit{},
+	}
+	for _, opt := range opts {
+		opt(&options)
+	}
+
+	pendingMetric := pendingHashReqs.WithLabelValues(string(hint))
+	pendingMetric.Add(float64(len(hashes)))
+
 	var eg errgroup.Group
 	var errs error
 	var mu sync.Mutex
-	for _, hash := range hashes {
+	for i, hash := range hashes {
+		if err := options.limiter.Acquire(ctx, 1); err != nil {
```

Comment:
looks like this is interruptible as well

Reply:
Yes `Acquire` blocks until a slot is available. `ctx` allows an early cancellation. I would also expect `Acquire` to return `ctx.Err()` as its error when the context is done.